src/java/org/apache/cassandra/utils/SortingIteratorNulls.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.utils;
+
+import java.lang.reflect.Array;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.function.Function;
+
+public class SortingIteratorNulls<T> implements Iterator<T>
```

Comment:
There are a couple more differences. After I saw an expected advantage for `LucenePriorityQueue` with slowed-down comparator, I made the second version to test the difference between two approaches:
- get rid of nulls, keep track of an active queue size, replace top item with another live item on removal;
- keep nulls in the list but compare them as bigger than live items, keep the size of the queue constant and replace top item with null on removal.

As described in the comment with the next commit, the latter turned out to be roughly as good (after some further optimizations) with fast comparators, and significantly better with slow ones, mostly because pushing down a null requires half the comparisons.

Reply:
The next commit selects `SortingIteratorNulls` as the one to go with, and removes the other implementations.